File: Sources/CountriesFlagsObjC/source/Countries/UIImage+Laos.m
//
//  UIImage+Laos.m
//  CountriesFlags
//

#import "Countries/UIImage+Laos.h"

@implementation UIImage (Laos)

+ (UIImage * _Nullable)flagLaosWithSize:(CGSize)size {
    UIColor* colorBlue = [UIColor colorWithRed:0.0f
                                         green:40.0f/255.0f
                                          blue:104.0f/255.0f
                                         alpha:1.0f];
    UIColor* colorRed = [UIColor colorWithRed:206.0f/255.0f
                                        green:17.0f/255.0f
                                         blue:38.0f/255.0f
                                        alpha:1.0f];
    UIColor* colorWhite = [UIColor whiteColor];
    
    UIGraphicsBeginImageContextWithOptions(size, NO, 0.0f);
    
    [colorRed setFill];
    UIRectFill(CGRectMake(0.0f, 0.0f, size.width, size.height));
    
    [colorBlue setFill];
    UIRectFill(CGRectMake(0.0f, size.height / 4.0f, size.width, size.height / 2.0f));
    
    [colorWhite setFill];
    CGFloat radius = size.height / 2.0f * (4.0f / 5.0f);
    [[UIBezierPath bezierPathWithOvalInRect:CGRectMake(size.width / 2.0f - radius / 2.0f,
                                                       size.height / 2.0f - radius / 2.0f,
                                                       radius,
                                                       radius)] fill];
    
    UIImage* image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return image;
}

@end

File: Sources/CountriesFlags/Countries/UIImage+Laos.swift
//
//  UIImage+Laos.swift
//  CountriesFlags
//

#if canImport(UIKit)

import UIKit

extension UIImage {
    /**
     [wikipedia]: https://en.wikipedia.org/wiki/Flag_of_Laos "wikipedia"
     
     Generates the flag of Laos.
     For more information see [wikipedia].
     
     - Parameters:
        - size: image size.
     
     - Returns: Returns the flag of Laos.
     */
    internal static func flagLaos(with size: CGSize) -> UIImage {
        let colorBlue = UIColor(red: 0.0,
                                green: 40.0 / 255.0,
                                blue: 104.0 / 255.0,
                                alpha: 1.0)
        let colorRed = UIColor(red: 206.0 / 255.0,
                               green: 17.0 / 255.0,
                               blue: 38.0 / 255.0,
                               alpha: 1.0)
        let colorWhite = UIColor.white

        UIGraphicsBeginImageContextWithOptions(size, true, 0.0)

        colorRed.setFill()
        UIRectFill(CGRect(origin: .zero, size: size))

        colorBlue.setFill()
        UIRectFill(CGRect(origin: CGPoint(x: 0.0, y: size.height / 4.0),
                          size: CGSize(width: size.width, height: size.height / 2.0)))

        colorWhite.setFill()
        UIBezierPath(arcCenter: size.midPoint,
                     radius: size.height / 2.0 * (4.0 / 5.0) / 2.0,
                     startAngle: 0.0,
                     endAngle: .pi * 2.0,
                     clockwise: true).fill()

        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()

        return image ?? UIImage()
    }
}

#endif
